proposals/TP1-xds-transport-next.md

```diff
@@ -0,0 +1,1028 @@
+  is at the discretion of the control plane(s).
+* `context parameters` are the URI query parameters and express contextual
+  information for selecting resource variants, for example `shard_id=1234&direction=inbound`.
+* A `processing directive` provides additional information to the client on how
+  the URI is to be interpreted. Two supported directives are:
+  * `alt=<xdstp:// URI>` expressing a fallback URI.
+  * `entry=<resource name>` providing an anchor referring to an inlined resource
+    for [list collections](#list). Resource names must be of the form `[a-zA-Z0-9_-\./]+`.
+
+Multiple directives are comma separated in the fragment component of the URI.
+
+An example application of the above URI scheme is:
+
+`xdstp://some.control.plane/envoy.config.route.v3.RouteConfiguration/foo/bar?shard_id=1234&direction=inbound`
+
+We distinguish between two forms of `xdstp://` URIs:
+
+* Uniform Resource Locators (URLs) instructing a client on how to locate a given
+  xDS resource. URLs may include processing directives.
+
+* Uniform Resource Names (URNs) referring to a specific xDS resource. URNs do not
+  include processing directives. `xdstp://` URNs provide self-contained cache keys.
+
+#### Proto3 expression
+
+`xdstp://` URIs are canonically encoded in `proto3` messages. The structure
+for this encoding is provided at https://github.com/cncf/xds/tree/main/xds/core/v3:
+
+* [URL](https://github.com/cncf/xds/blob/main/xds/core/v3/resource_locator.proto) definition
+* [URN](https://github.com/cncf/xds/blob/main/xds/core/v3/resource_name.proto) definition
+
+#### Normalization
+
+Two `xdstp://` URNs are considered equivalent if they match component-wise
+modulo context parameter ordering.
+
+## Context parameters
+
+Context parameters in URNs presented by the client to the server will be dervied
+by the following process. Using an example of a URL
+`xdstp://some-authority/some.type/foo?xds.bar=baz`:
+
+1. Static `Node`-derived context parameters are first added. These are prefixed with
+   `xds.node.`. The set of `Node`-derived context parameters is specified in the
+   bootstrap on a per-resource type basis. The key reflects the `Node` proto3
+   field structure, e.g. `xds.node.locality.sub_zone=some_sub_zone`,
+   `xds.node.metadata.bar="a"`, `xds.node.user_agent_version=1.2.3`.
+
+2. Dynamic context parameters are added next. These are free form and at the
+   discretion of the client. They can provide client-specific fields such as
+   `xds.shard_id=1234` and potentially change through the lifetime of an xDS
+   stream.
+
+3. Context parameters from the URL are added next, in the above example
+   `xds.bar=baz`.
+
+4. [Client feature
+   capability](https://www.envoyproxy.io/docs/envoy/latest/api/client_features)
+   context parameters are then added, e.g. clusters at the client might have
+   `{xds.client_feature.lb.least_loaded: true}`. Client feature capabilities
```

Comment:
I think we had also said that these might be a little bit viral, because some resources can be sent as part of another resource.  For example, the RouteConfig may be inlined into the Listener resource or it may be obtained separately, and the client doesn't know when it sends the LDS request whether the response will include the RouteConfig.  This means that if the client has a feature that affects the RouteConfig, it needs to include that in all LDS or RDS requests it sends.

Reply:
Thanks for reminding me; I've updated the text to explain this.

Reply:
While I've updated it, I'm actually wondering if this is getting a bit impractical; if you have something like filters, which can be literally _anything_, this seems a lot of client features to be sending. E.g. We need to send client features for the Dubbo filter references, as well as those for a RouteConfiguration embedded in HCM. 

Relying on the user agent (name+version) and having some well known map from that to client features starts to sound attractive.